### PR TITLE
refactor: data config

### LINKED ISSUE #161: Refactor data configuration
@sehilyi , I am looking at the data interface and wondering your opinion about the following refactoring

``` javascript
interface BasicSingleTrack extends CommonTrackDef {
    data: DataDeep;
    ....
}


interface DataDeep {
    dataSource: DataSource
    dataConfig: DataConfig
}

interface DataSource {
    type: "json"|"csv"|"vector"|"bed"|"multivec",
    url: string 
}

interface DataConfig = JSONDataConfig | CSV DataConfig | VectorDataConfig | MultivecDataConfig | BEDDataConfig;

interface JSONDataConfig {
    values: Datum[] 
}

interface CSVDataConfig {
    separator?: string;
    headerNames?: string[];
    chromosomePrefix?: string;
    longToWideId?: string;
}

interface BEDDataConfig {
    genomicFields: { index: number; name: string }[];
    valueFields?: { index: number; name: string; type: 'nominal' | 'quantitative' }[];
    exonIntervalFields?: [{ index: number; name: string }, { index: number; name: string }];
}

interface MultivecDataConfig {
    column: string;
    row: string;
    value: string;
    categories?: string[];
    start?: string;
    end?: string;
    bin?: number;
}

interface VectorDataConfig {
    column: string;
    value: string;
    start?: string;
    end?: string;
    bin?: number; 
}


```

### PR DESCRIPTION
**Pull request by wangqianwen0418**
Fix #161

### CODE REVIEW

**Comment by sehilyi** [comment]
@wangqianwen0418 I am not sure if this is a draft or ready to be reviewed. Can you assign me as a reviewer when you feel your PR is ready?

**Comment by sehilyi** [comment]
Also, just an FYI, you can put "Fix" in front of the issue number (i.e., "#161") if you feel the issue is being addressed by this PR. The predefined string ("Fix") will automatically close the mapped issue(s) whenever this PR is merged to the master branch.

**Comment by wangqianwen0418** [comment]
sorry for the unclearness caused!
Yes, it is ready and please have a review

File: src/core/gosling.schema.ts [hunk 1]
```diff
@@ -170,7 +170,6 @@ export interface CommonTrackDef {
 export interface DataTrack extends CommonTrackDef {
     // !!! The below properties should be required ones since metadata determines the visualization type.
```
Comment (sehilyi) [inline src/core/gosling.schema.ts]:
```suggestion
```

File: src/core/gosling.schema.ts [hunk 2]
```diff
@@ -63,45 +54,54 @@ export interface DataDeepGoslingCommon {
...
 
-export interface CSVDataGosling extends DataDeepGoslingCommon {
+export interface Datum {
+    [k: string]: number | string;
+}
+
+export interface CSVData {
     type: 'csv';
-    url?: string;
+    url: string;
```
Comment (sehilyi) [inline src/core/gosling.schema.ts:63]: Thank you for correcting this.

File: src/core/utils/spec-preprocess.ts [hunk 3]
```diff
@@ -136,7 +136,7 @@ export function getMultivecTemplate(
  */
 export function overrideTemplates(spec: GoslingSpec) {
     spec.tracks.forEach((t, i) => {
-        if (!t.metadata || !IsDataMetadata(t.metadata)) {
+        if (!t.data || !IsDataDeepTileset(t.data)) {
             // if `metadata` is not specified, we can not provide a correct template since we do not know the exact data type.
```
Comment (sehilyi) [inline src/core/utils/spec-preprocess.ts]:
```suggestion
            // if `data` is not specified, we can not provide a correct template.
```

File: src/gosling-track/gosling-track.ts [hunk 4]
```diff
@@ -420,7 +420,7 @@ function GoslingTrack(HGC: any, ...args: any[]): any {
                             const { oneOf } = filter;
                             tile.tileData.tabularDataFiltered = tile.tileData.tabularDataFiltered.filter(
                                 (d: { [k: string]: number | string }) => {
-                                    return not
+                                    return not ?? false
```
Comment (sehilyi) [inline src/gosling-track/gosling-track.ts]:
```suggestion
                                    return not
```

File: src/gosling-track/gosling-track.ts [hunk 5]
```diff
@@ -429,7 +429,7 @@ function GoslingTrack(HGC: any, ...args: any[]): any {
                             const { inRange } = filter;
                             tile.tileData.tabularDataFiltered = tile.tileData.tabularDataFiltered.filter(
                                 (d: { [k: string]: number | string }) => {
-                                    return not
+                                    return not ?? false
```
Comment (sehilyi) [inline src/gosling-track/gosling-track.ts]:
```suggestion
                                    return not
```

File: src/gosling-track/gosling-track.ts [hunk 6]
```diff
@@ -438,7 +438,9 @@ function GoslingTrack(HGC: any, ...args: any[]): any {
                             const { include } = filter;
                             tile.tileData.tabularDataFiltered = tile.tileData.tabularDataFiltered.filter(
                                 (d: { [k: string]: number | string }) => {
-                                    return not ? `${d[field]}`.includes(include) : !`${d[field]}`.includes(include);
+                                    return not ?? false
```
Comment (sehilyi) [inline src/gosling-track/gosling-track.ts]:
```suggestion
                                    return not
```

File: src/core/utils/superpose.test.ts [hunk 7]
```diff
@@ -62,20 +62,4 @@ describe('Spread Tracks By Data', () => {
         expect('data' in spread[1]).toBe(true); // Any superposed tracks w/ data/metadata will be spread
         expect(spread[1].superposeOnPreviousTrack).toBe(true); // Any spread tracks will have `superposeOnPreviousTrack` flags
     });
```
Comment (sehilyi) [inline src/core/utils/superpose.test.ts:64]:
```suggestion
    });
    it('superpose: [{}, { data }, { data }]', () => {	
        const spread = spreadTracksByData([	
            {	
                superpose: [	
                    {},	
                    { data: { type: 'csv', url: '' } },	
                    { data: { type: 'vector', url: '', column: 'c', value: 'p' } }	
                ]	
            }	
        ]);	
        expect(spread).toHaveLength(3);	
        expect('data' in spread[1]).toBe(true);	
        expect('data' in spread[2]).toBe(true);	
        expect(spread[1].superposeOnPreviousTrack).toBe(true); // Any spread tracks will have `superposeOnPreviousTrack` flags	
        expect(spread[2].superposeOnPreviousTrack).toBe(true); // Any spread tracks will have `superposeOnPreviousTrack` flags	
    });
```

I am unsure if this test has been removed by mistake, but we shouldn't be removing tests.

File: src/gosling-track/gosling-track.ts (hunk 4 above)
Comment (sehilyi) [reply src/gosling-track/gosling-track.ts]: I think this change is from your previous PR on the same branch, which is then fixed before merging the PR.

**Review by sehilyi — CHANGES_REQUESTED** [review]
Thank you for proposing and working on this! 

I think this is a way better than before in terms of the simplicity/readability (e.g., good to remove needless `tileset` strings).

I've left one suggestion on putting back a test along with some minor updates on comment statements. Other than that, it looks good to go.

**Review by sehilyi — APPROVED** [review]

**Pull request merged**